singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,374 @@
+    return;
+  }
+
+  for (ind = 0; ind < 39; ind++)
+    _d2to40[ind] = _expconsts[ind];
+
+  _VIP[0] = _B0 / _rho0 + pow(_A0 * _B0 / _rho0, 2.0) * _T0 / _Cv0; // sound speed squared
+  _VIP[1] = 3.0 / 2.0 * (_BP0 - 1.0);                               // exponent
+  _VIP[2] = 1.0;                                                    // prefactor f0
+  _VIP[3] = 0.0;                                                    // prefactor f1
+
+  for (ind = 4; ind < 43; ind++)
+    _VIP[ind] = _d2to40[ind - 4];
+  ind = 42;
+  while (_VIP[ind] == 0.0)
+    ind--;
+  while (ind >= 4) {
+    ind--;
+    _VIP[ind] = _VIP[ind] - (ind - 1) / _VIP[1] * _VIP[ind + 1];
+  }
+}
+
+PORTABLE_INLINE_FUNCTION void Vinet::Vinet_F_DT_func(const Real rho, const Real T,
+                                                     Real *output) const {
+  int pref0vp = -2, pref0vip = 2, ind;
+  Real x, x2inv, onemx, etatimes1mx, expetatimes1mx, temp;
+  Real sumP = 0.0, sumB = 0.0, sumE = 0.0;
+  Real entropy, energy, pressure, dpdrho, dpdt, dedt, dedrho, soundspeed;
+
+  ind = 40;
+  while (_VIP[pref0vip + ind] == 0.0)
+    ind--;
+
+  x = pow(_rho0 / rho, 1.0 / 3.0); /*rho-dependent*/
+  x2inv = 1.0 / pow(x, 2.0);
+  onemx = 1.0 - x;
+  etatimes1mx = _VIP[1] * onemx;
+  expetatimes1mx = exp(etatimes1mx);
+
+  while (ind >= 2) {
+    sumP = _d2to40[pref0vp + ind] + onemx * sumP;
+    sumB = _d2to40[pref0vp + ind] * ind + onemx * sumB;
+    sumE = _VIP[pref0vip + ind] + onemx * sumE;
+    ind--;
+  }
+  while (ind >= 0) {
+    sumE = _VIP[pref0vip + ind] + onemx * sumE;
+    ind--;
+  }
+  sumP = 1.0 + sumP * pow(onemx, 2.0);
+  sumB = sumB * onemx;
+  sumE = _VIP[1] * onemx * sumE;
+
+  /* T0 isotherm */
+  energy = 9.0 * _B0 / pow(_VIP[1], 2.0) / _rho0 *
+               (_VIP[pref0vip] - expetatimes1mx * (_VIP[pref0vip] - sumE)) -
+           _A0 * _B0 * _T0 * (1.0 / _rho0 - 1.0 / rho);
+  pressure = 3.0 * _B0 * x2inv * onemx * expetatimes1mx * sumP;
+  temp = (1.0 + onemx * (_VIP[1] * x + 1.0)) * sumP + x * onemx * sumB;
+  temp = _B0 / _rho0 * x * expetatimes1mx * temp;
+
+  /* Go to required temperature */
+  energy = energy + _Cv0 * (T - _T0) + _E0;
+  pressure = pressure + _A0 * _B0 * (T - _T0);
+  dpdrho = temp;
+  dpdt = _A0 * _B0;
+  dedt = _Cv0;
+  dedrho = (pressure - T * _A0 * _B0) / pow(rho, 2.0);
+  if (T < 0.0) {
+    printf("warning!! negative temperature");
+    entropy = 0.0;
+  } else
+    entropy = _A0 * _B0 * (1.0 / rho - 1 / _rho0) + _Cv0 * log(T / _T0) + _S0;
+  soundspeed = std::max(0.0, dpdrho + T * pow(dpdt / rho, 2.0) / _Cv0);
+  soundspeed = sqrt(soundspeed);
+
+  output[0] = energy;
+  output[1] = pressure;
+  output[2] = dpdrho;
+  output[3] = dpdt;
+  output[4] = dedt;
+  output[5] = dedrho;
+  output[6] = entropy;
+  output[7] = soundspeed;
+
+  return;
+}
+PORTABLE_INLINE_FUNCTION Real Vinet::InternalEnergyFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real output[8];
+  Vinet_F_DT_func(rho, temp, output);
```

Comment:
Do you need to call the `Vinet_F_DT_func` and calculate everything for every variable call? This seems hugely wasteful.

Reply:
I'm not going to insist on this change. Given the way your code works, it would be  a big change. But we should note somewhere that calling the per-quantity calls for `Vinet` is slow compared to calling `FIllEos` because the per quantity calls compute the entire thermodynamic state.